app/src/modules/environmental/environmental.c

```diff
@@ -26,8 +28,100 @@ BUILD_ASSERT(CONFIG_APP_ENVIRONMENTAL_WATCHDOG_TIMEOUT_SECONDS >
 			"Watchdog timeout must be greater than maximum execution time");
 
 static const struct device *const sensor_dev = DEVICE_DT_GET(DT_ALIAS(gas_sensor));
-static bool time_available;
 
+/* Forward declarations */
+static struct s_object s_obj;
+static void sample(void);
+
+/* State machine */
+
+/* Defininig the module states.
+ *
+ * STATE_INIT: The environmental module is initializing and waiting for time to be available.
+ * STATE_SAMPLING: The environmental module is ready to sample upon receiving a trigger.
+ */
+enum environmental_module_state {
+	STATE_INIT,
+	STATE_SAMPLING,
+};
+
+/* User defined state object.
+ * Used to transfer data between state changes.
+ */
+struct s_object {
+	/* This must be first */
+	struct smf_ctx ctx;
+
+	/* Last channel type that a message was received on */
+	const struct zbus_channel *chan;
+
+	/* Last received message */
+	uint8_t *msg;
+};
+
+/* Forward declarations of state handlers */
+static void state_init_run(void *o);
+static void state_sampling_run(void *o);
+
+static struct s_object s_obj;
+static const struct smf_state states[] = {
+	[STATE_INIT] =
+		SMF_CREATE_STATE(NULL, state_init_run, NULL,
+				 NULL,	/* No parent state */
+				 NULL), /* No initial transition */
+	[STATE_SAMPLING] =
+		SMF_CREATE_STATE(NULL, state_sampling_run, NULL,
+				 NULL,
+				 NULL),
+};
+
+/* State handlers */
+
+static void state_init_run(void *o)
+{
+	int err;
+	struct s_object *state_object = o;
+
+	if (&TIME_CHAN == state_object->chan) {
+		enum time_status time_status;
+
+		err = zbus_chan_read(&TIME_CHAN, &time_status, K_FOREVER);
+		if (err) {
+			LOG_ERR("zbus_chan_read, error: %d", err);
+			return;
+		}
+
+		if (time_status == TIME_AVAILABLE) {
+			LOG_DBG("Time available, sampling can start");
+
+			STATE_SET(STATE_SAMPLING);
+		}
+	}
+}
+
+static void state_sampling_run(void *o)
+{
+	struct s_object *state_object = o;
+
+	if (&TRIGGER_CHAN == state_object->chan) {
+		int err;
+		enum trigger_type trigger_type;
+
+		err = zbus_chan_read(&TRIGGER_CHAN, &trigger_type, K_FOREVER);
```

Comment:
Won't `K_FOREVER` prevent the watchdog to be fed in the `battery_task`? Same with the `TIME_CHAN` above.